WordPress/gutenberg
# Interactivity API: Use modules instead of scripts in the frontend

## What?


Built on top of https://github.com/WordPress/gutenberg/pull/56118.

This is the third of a series of experiments to understand the requirements of a hypothetical Modules API. More info in the Tracking Issue:

- https://github.com/WordPress/gutenberg/issues/55942

This experiment aims to implement an API that is more complex than the ones implemented in the [first](https://github.com/WordPress/gutenberg/pull/56092) and [second](https://github.com/WordPress/gutenberg/pull/56118) experiments, adding a required server dependency graph to be able to print only the necessary modules in the import map, apart from preloading modules to avoid initial waterfalls:

- An API to register modules using:
  - A module identifier (required)
  - The full URL or the path (required)
  - The static and dynamic dependencies (required)
  - A version (optional)
- An API to enqueue modules.
- Only the modules needed for the enqueued modules have an entry in the import map.
- Enqueued modules add a `<script>` tag in the head.
- The static dependencies are preloaded when the module is enqueued using a `<link rel="modulepreload">` tag in the head.

## Why?


In the [first experiment](https://github.com/WordPress/gutenberg/pull/56092), it's not easy to know which modules need to be preloaded and, therefore, it's not easy to avoid the initial waterfalls.

In the [second experiment](https://github.com/WordPress/gutenberg/pull/56118), all the registered modules are added to the import map, even the ones that are not used on the page. In this version, declaring the dependencies is required, so we can know all the modules that are going to be used on the page and only include those in the import map.

## How?


By leveraging a new `Gutenberg_Modules` class that acts as a singleton and a couple of functions to expose the functionality:

- `gutenberg_register_module` to register modules and add entries to the import map.
- `gutenberg_enqueue_module` to enqueue modules.

Then, we print three times using the `wp_head` action:

- The import map.
- The enqueued modules.
- The preloaded modules.

## Testing Instructions





- Add some interactive blocks: Navigation, Query or Image block.
- Check that in the frontend, there is an import map script.
- Check that scripts with `type="module"` are loaded by the interactive blocks.
- Check that the `@wordpress/interactivity` module is loaded by the other modules using `import ... from "@wordpress/interactivity"`.
- Check that the `@wordpress/interactivity` module is downloaded at the same time as the rest (instead of after, like in the [first experiment](https://github.com/WordPress/gutenberg/pull/56092)), and that there is a link tag with `rel="modulepreload"` in the head.
- Check that the import map only has the `@wordpress/interactivity` module.


## Follow-up tasks

- [ ] Replace the syntax of the dependencies array with [this one](https://github.com/WordPress/gutenberg/pull/56143#discussion_r1395753721).
- [ ] Filter enqueued modules when printing the preloaded modules.
